src/hotspot/share/nmt/memLogRecorder.cpp

```diff
@@ -0,0 +1,839 @@
+  }
+}
+
+void NMT_LogRecorder::finish() {
+  if (!NMT_MemoryLogRecorder::instance()->done()) {
+    NMT_MemoryLogRecorder::instance()->finish();
+  }
+  if (!NMT_VirtualMemoryLogRecorder::instance()->done()) {
+    NMT_VirtualMemoryLogRecorder::instance()->finish();
+  }
+}
+
+void NMT_LogRecorder::replay() {
+  char* NMTBenchmarkRecordedPID = getenv("NMTBenchmarkRecordedPID");
+  if (NMTBenchmarkRecordedPID != nullptr) {
+    int pid = atoi(NMTBenchmarkRecordedPID);
+    NMT_MemoryLogRecorder::instance()->replay(pid);
+    NMT_VirtualMemoryLogRecorder::instance()->replay(pid);
+    os::exit(0);
+  }
+}
+
+void NMT_LogRecorder::init() {
+  _threads_names_size = 1;
+  _threads_names = (thread_name_info*)raw_realloc(_threads_names, _threads_names_size*sizeof(thread_name_info));
+  _done = true;
+  _count = 0;
+}
+
+void NMT_LogRecorder::get_thread_name(char* buf) {
+#if defined(__APPLE__)
+  if (pthread_main_np()) {
+    strcpy(buf, "main");
+  } else {
+    if (Thread::current() != nullptr) {
+      strncpy(buf, Thread::current()->name(), MAXTHREADNAMESIZE-1);
+      // filter out temp thread names
+      if (strcmp(buf, "Unknown thread") == 0) {
+        buf[0] = 0;
+      }
+    }
+  }
+#elif defined(LINUX)
+  pthread_getname_np(pthread_self(), buf, MAXTHREADNAMESIZE);
+#elif defined(_WIN64)
+  // TODO: NMT_LogRecorder::thread_name
+#endif
+}
+
+// first time we see a new thread id, we add it
+// second time we see a thread we get its name
+void NMT_LogRecorder::logThreadName() {
+  {
+    bool found = false;
+    long int tid = os::current_thread_id();
+    for (size_t i = 0; i < _threads_names_size; i++) {
+      if (_threads_names[i].thread == tid) {
+        // we found the entry in the table
+        found = true;
+        if (_threads_names[i].name[0] == 0) {
+          // but we don't have the name yet, so try
+          NMT_LogRecorder::get_thread_name(_threads_names[i].name);
+          //tty->print(" got name for thread %6ld:%lx [%s]\n", tid, tid, _threads_names[i].name);
+        }
+        break;
+      }
+    }
+    if (!found) {
+      // tty->print(" added:%6ld:%lx [%6zu]\n", tid, tid, _threads_names_size);
+      size_t i = _threads_names_size-1;
+      _threads_names[i].thread = tid;
+      _threads_names[i].name[0] = 0;
+      _threads_names_size++;
+      _threads_names = (thread_name_info*)raw_realloc(_threads_names, _threads_names_size*sizeof(thread_name_info));
+    }
+  }
+}
+
+#define IS_FREE(e)           ((e->requested == 0) && (e->old == nullptr))
+#define IS_REALLOC(e)        ((e->requested  > 0) && (e->old != nullptr))
+#define IS_MALLOC(e)         ((e->requested  > 0) && (e->old == nullptr))
+
+#define ALLOCS_LOG_FILE "hs_nmt_pid%p_allocs_record.log"
+#define THREADS_LOG_FILE "hs_nmt_pid%p_threads_record.log"
+#define INFO_LOG_FILE "hs_nmt_pid%p_info_record.log"
+#define BENCHMARK_LOG_FILE "hs_nmt_pid%p_benchmark.log"
+#define VALLOCS_LOG_FILE "hs_nmt_pid%p_virtual_allocs_record.log"
+
```

Comment:
Please replace all `#define`s with constants instead, do not use `ALL_CAPS` formatting for the replacements.